s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/output/OutputWriter.java

```diff
@@ -0,0 +1,93 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.connect.s3.source.output;
+
+import static io.aiven.kafka.connect.s3.source.S3SourceTask.BUCKET;
+import static io.aiven.kafka.connect.s3.source.S3SourceTask.PARTITION;
+import static io.aiven.kafka.connect.s3.source.S3SourceTask.TOPIC;
+import static io.aiven.kafka.connect.s3.source.config.S3SourceConfig.SCHEMA_REGISTRY_URL;
+import static io.aiven.kafka.connect.s3.source.config.S3SourceConfig.VALUE_SERIALIZER;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.lang.reflect.InvocationTargetException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+
+import io.aiven.kafka.connect.s3.source.config.S3SourceConfig;
+import io.aiven.kafka.connect.s3.source.utils.OffsetManager;
+
+import io.confluent.kafka.serializers.KafkaAvroSerializer;
+import org.apache.avro.generic.GenericRecord;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public interface OutputWriter {
+
+    Logger LOGGER = LoggerFactory.getLogger(AvroWriter.class);
+
+    void configureValueConverter(Map<String, String> config, S3SourceConfig s3SourceConfig);
+    void handleValueData(Optional<byte[]> optionalKeyBytes, InputStream inputStream, String topic,
+            List<ConsumerRecord<byte[], byte[]>> consumerRecordList, S3SourceConfig s3SourceConfig, int topicPartition,
+            long startOffset, OffsetManager offsetManager, Map<Map<String, Object>, Long> currentOffsets);
+
+    default ConsumerRecord<byte[], byte[]> getConsumerRecord(final Optional<byte[]> key, final byte[] value,
+            final String bucketName, final String topic, final int topicPartition, final OffsetManager offsetManager,
+            final Map<Map<String, Object>, Long> currentOffsets, final long startOffset) {
+        final Map<String, Object> partitionMap = new HashMap<>();
+        partitionMap.put(BUCKET, bucketName);
+        partitionMap.put(TOPIC, topic);
+        partitionMap.put(PARTITION, topicPartition);
+
+        long currentOffset;
+
+        if (offsetManager.getOffsets().containsKey(partitionMap)) {
+            currentOffset = offsetManager.incrementAndUpdateOffsetMap(partitionMap);
+        } else {
+            currentOffset = currentOffsets.getOrDefault(partitionMap, startOffset);
+            currentOffsets.put(partitionMap, currentOffset + 1);
+        }
+
+        return new ConsumerRecord<>(topic, topicPartition, currentOffset, key.orElse(null), value);
+    }
+
+    default byte[] serializeAvroRecordToBytes(final List<GenericRecord> avroRecords, final String topic,
```

Comment:
As this is specific avro serialization I htink it would be better placed in AvroWriter.java

Reply:
This piece of code is for both parquet and avro, so put it at interface. But now moved them to Utils class.

Reply:
Refactored.